Mathlib/CategoryTheory/Bicategory/Kan/HasKan.lean

```diff
@@ -0,0 +1,308 @@
+
+## TODO
+
+* `ran f g` is the right Kan extension of `g` along `f`, and is denoted by `f⁺⁺ g`.
+* `ranLift f g` is the right Kan lift of `g` along `f`, and is denoted by `f₊₊ g`.
+
+-/
+
+noncomputable section
+
+namespace CategoryTheory
+
+namespace Bicategory
+
+universe w v u
+
+variable {B : Type u} [Bicategory.{w, v} B] {a b c : B}
+
+open Limits
+
+section LeftKan
+
+open LeftExtension
+
+variable {f : a ⟶ b} {g : a ⟶ c}
+
+/-- The existence of a left Kan extension of `g` along `f`. -/
+class HasLeftKanExtension (f : a ⟶ b) (g : a ⟶ c) : Prop where
+  hasInitial : HasInitial <| LeftExtension f g
+
+theorem LeftExtension.IsKan.hasLeftKanExtension {t : LeftExtension f g} (H : IsKan t) :
+    HasLeftKanExtension f g :=
+  ⟨IsInitial.hasInitial H⟩
+
+instance [HasLeftKanExtension f g] : HasInitial <| LeftExtension f g :=
+  HasLeftKanExtension.hasInitial
+
+/-- The left Kan extension of `g` along `f` at the level of structured arrows. -/
+def lanLeftExtension (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : LeftExtension f g :=
+  ⊥_ (LeftExtension f g)
+
+/-- The left Kan extension of `g` along `f`. -/
+def lan (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : b ⟶ c :=
+  (lanLeftExtension f g).extension
+
+/-- `f⁺ g` is the left Kan extension of `g` along `f`.
+```
+  b
+  △ \
+  |   \ f⁺ g
+f |     \
+  |       ◿
+  a - - - ▷ c
+      g
+```
+-/
+scoped infixr:90 "⁺ " => lan
+
+@[simp]
+theorem lanLeftExtension_extension (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] :
+    (lanLeftExtension f g).extension = f⁺ g := rfl
+
+/-- The unit for the left Kan extension `f⁺ g`. -/
+def lanUnit (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : g ⟶ f ≫ f⁺ g :=
+  (lanLeftExtension f g).unit
+
+@[simp]
+theorem lanLeftExtension_unit (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] :
+    (lanLeftExtension f g).unit = lanUnit f g := rfl
+
+/-- Evidence that the `Lan.extension f g` is a Kan extension. -/
+def lanIsKan (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : (lanLeftExtension f g).IsKan :=
+  initialIsInitial
+
+variable {f : a ⟶ b} {g : a ⟶ c}
+
+/-- The family of 2-morphisms out of the left Kan extension `f⁺ g`. -/
+def lanDesc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    f⁺ g ⟶ s.extension :=
+  (lanIsKan f g).desc s
+
+@[reassoc (attr := simp)]
+theorem lanUnit_desc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    lanUnit f g ≫ f ◁ lanDesc s = s.unit :=
+  (lanIsKan f g).fac s
+
+@[simp]
+theorem lanIsKan_desc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    (lanIsKan f g).desc s = lanDesc s :=
+  rfl
+
+theorem Lan.existsUnique [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    ∃! τ, lanUnit f g ≫ f ◁ τ = s.unit :=
+  (lanIsKan f g).existsUnique _
+
+/-- We say that a 1-morphism `h` commutes with the left Kan extension `f⁺ g` if the whiskered
+left extension for `f⁺ g` by `h` is a Kan extension of `g ≫ h` along `f`. -/
+class Lan.CommuteWith
```

Comment:
In the API for `Lan.CommuteWith`, I think that the lemmas should be stated not only for the chosen `lanLeftExtension f g`, but all for any left extension which satisfies `IsKan`. In particular, I think that the `Lan.CommuteWith.isKan` definition should be generalized as `LeftExtension.IsKan.whiskerOfCommute`.

Reply:
Sorry for the late reply. 

I may not understand the point yet. Could you explain in a little more detail your idea about `LeftExtension.IsKan.whiskerOfCommute`?

Reply:
In addition to `isKan`, we would have:
```lean
def _root_.CategoryTheory.Bicategory.LeftExtension.IsKan.whiskerOfCommute
    {f : a ⟶ b} {g : a ⟶ c} {t : LeftExtension f g} (ht : t.IsKan) {x : B} (h : c ⟶ x)
    [HasLeftKanExtension f g] [Lan.CommuteWith f g h] :
    (t.whisker h).IsKan := by
  sorry
```
Of course, `t` must be isomorphic to `lanLeftExtension`, so that this is true.

Basically, any definition/lemma where `lanLeftExtension` appears should have a version which applies to any left extension `t` which satisfies `t.IsKan`.

Then, I also think that it would be better if `CommuteWith` was like:
```lean
class Lan.CommuteWith (f : a ⟶ b) (g : a ⟶ c) {x : B} (h : c ⟶ x) : Prop :=
  hasLeftKanExtension : HasLeftKanExtension f g := by infer_instance
  commute : Nonempty <| IsKan <| (lanLeftExtension f g).whisker h
```
Also, we should have a constructor for `CommuteWith` which takes as input a `t` as above such that `t.IsKan` and `(t.whisker h).IsKan`.

Reply:
@joelriou Thank you for your explanation. I have added lemmas. My feeling is that these are problems of transporting `IsKan` along an isomorphism.